SwiftCompilerSources/Sources/Optimizer/InstructionSimplification/SimplifyBeginBorrow.swift

```diff
@@ -118,14 +118,28 @@ private extension Value {
   }
 
   func isDestroyed(after nonDestroyUser: Instruction) -> Bool {
-    uses.getSingleUser(notOfType: DestroyValueInst.self) == nonDestroyUser
+    return uses.getSingleUser(notOfType: DestroyValueInst.self) == nonDestroyUser &&
+           nonDestroyUser.dominates(destroysOf: self)
   }
 
   func replaceAllDestroys(with replacement: Value, _ context: SimplifyContext) {
     uses.filterUsers(ofType: DestroyValueInst.self).replaceAll(with: replacement, context)
   }
 }
 
+private extension Instruction {
+  func dominates(destroysOf value: Value) -> Bool {
```

Comment:
For the record. Anyone who sees `instruction.dominates` will assume it is an implementation of general instruction dominance. An extension of a fundamental type should have an implementation that makes sense wherever that type may be used, regardless of whether the extension is private.